sycl/doc/extensions/supported/sycl_ext_oneapi_backend_level_zero.md

```diff
@@ -349,6 +398,16 @@ Applications must make sure that the Level-Zero handles themselves aren't used s
 Practically speaking, and taking into account that SYCL runtime takes ownership of the Level-Zero handles,
 the application should not attempt further direct use of those handles.
 
+### 4.5 Interoperability buffer synchronization rules
+
+A SYCL buffer that is constructed with this interop API uses the Level Zero memory buffer for its full lifetime, and the contents of the Level Zero memory buffer are unspecified for the lifetime of the SYCL buffer. If the application modifies the contents of that Level Zero memory buffer during the lifetime of the SYCL buffer, the behavior is undefined. The initial contents of the SYCL buffer will be the initial contents of the Level Zero memory buffer at the time of the SYCL buffer's construction.
+
+The behavior of the SYCL buffer destructor depends on the Ownership flag. As with other SYCL buffers, this behavior is triggered only when the last reference count to the buffer is dropped, as described in the core SYCL specification section 4.7.2.3, "Buffer synchronization rules".
+
+* If the ownership is keep (i.e. the application retains ownership of the Level Zero memory buffer), then the SYCL buffer destructor blocks until all work in queues on the buffer have completed. The buffer's contents are written to the Level Zero memory buffer by the time the destructor completes.

```

Comment:
I don't think we should do that because 1) it would be blocking the host when no one is possibly needing this and 2) we don't know where to write the final bits to (remember there can be multiple underlying allocations for multiple device in the context). Instead, I suggest that we say that anyone wanting to grab the SYCL's buffer data should use `get_native()` (and that one will return all valid bits).

Reply:
What if someone accessed interop buffer on host using host accessor and changed it. Which native memory handle we will return in this case? I believe, we can't just return pointer to the host buffer and we will need to do some copy.

Another case would be when interop buffer is used in different contexts.  Let's say interop buffer is created using Context1 and we access this buffer in a kernel submitted to the queue in Context1.
But later we access this buffer in different context Context2 (i.e. in the kernel which submitted to the queue in Context2).
Scheduler will create a separate pi_mem instance for the second context and will copy data from one pi_mem to another through host.

In the current version of the patch we copy the latest contents of the buffer to the original memory allocation in Context1 if user provided "keep" ownership (during buffer destruction).

Which memory handle should get_native return in this case? If we return memory handle from Context2 (which has the latest data) then we will need somehow to get info about the most up-to-date isntance of the pi_mem object from the scheduler, because only scheduler has this info. sycl::buffer references only original pi_mem instance which was created in the constructor.

Reply:
SYCL RT already does proper sync/copies for cross-context SYCL buffer accesses and those in host task. So nothing is special for interop, no?

> Which memory handle should get_native return in this case?

We will return any allocation in the context of the buffer, which is "Context1" in your example. Again, SYCL RT should always sync/copy data to "Context1", after it was changed in "Context2", I think.

Reply:
> We will return any allocation in the context of the buffer, which is "Context1" in your example. Again, SYCL RT should always sync/copy data to "Context1", after it was changed in "Context2", I think.
Hm, no, if the Context2 is the last context where we worked with a buffer then up-to-date data will be in Context2, it will not be automatically copied back to Context1.

Reply:
This was idea behind introduced BaseT::updateInteropMemory(); - to copy data back to original context.

Reply:
> if the Context2 is the last context where we worked with a buffer then up-to-date data will be in Context2, it will not be automatically copied back to Context1.

But doing `Buffer.get_native()` is essentially "working" with the buffer in its original context. Why would that not cause a copy back?

Reply:
I thought you mean that scheduler copies back by default in all cases.
If we implement get_native() to copy data back then of course it will copy back.

Reply:
I think we have to have it implemented this way otherwise data returned in get_native() pointer is invalid an hence unusable. And if we have it right, then I don't see why we need to do so in destructor.

Reply:
I'm concerned about the spec change in 2a21b302616bc0c1d6c129107ef69d4498c63256.  A common idiom in SYCL is like this:

```
void foo(sycl::queue q) {
  int *mem = malloc(n * sizeof(int));
  {
    sycl::buffer buf{mem, {N}};
    q.submit([&](sycl::handler cgh) {
      sycl::accessor acc{buf, cgh};
      cgh.parallel_for({N}, [=](sycl::id i) {/*...*/});
    });
    // buffer goes out of scope, destructor blocks until kernel completes and data written back to "mem"
  }
  /* ... */ = mem[0];  // Read data computed from kernel
}
```

I think it will be confusing if buffers created from interop behave differently.  For example, someone might write this code:

```
void foo(sycl::queue q) {
  // Create a USM shared memory allocation in "mem".
  int *mem;
  zeMemAllocShared(/*...*/);
  {
    sycl::buffer buf = sycl::make_buffer<backend::ext_oneapi_level_zero, int>(
        {mem, ext::oneapi::level_zero::ownership::keep}, q.get_context());
    q.submit([&](sycl::handler cgh) {
      sycl::accessor acc{buf, cgh};
      cgh.parallel_for({N}, [=](sycl::id i) {/*...*/});
    });
    // buffer goes out of scope, but destructor dos NOT block
  }
  /* ... */ = mem[0];  // This is unsafe because kernel may not have completed yet.
}
```

Reply:
Thank you, after getting some info from customer it looks like we need to copy back and block if user preferred to keep ownership of the memory handle. So returned the code to do copy back in case of "keep".

Reply:
I still see this in https://github.com/intel/llvm/blob/3cd57ed11e9a9529b37f6ad822f3f7bfa5ef2ed1/sycl/doc/extensions/supported/sycl_ext_oneapi_backend_level_zero.md

> The SYCL buffer destructor does not need to block even if work on the buffer has not completed.

Reply:
You are right, sorry, fixed this part.